presto-main/src/main/java/com/facebook/presto/execution/scheduler/nodeSelection/SimpleNodeSelector.java

```diff
@@ -111,19 +114,22 @@ public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTas
         NodeMap nodeMap = this.nodeMap.get().get();
         NodeAssignmentStats assignmentStats = new NodeAssignmentStats(nodeTaskMap, nodeMap, existingTasks);
 
-        ResettableRandomizedIterator<InternalNode> randomCandidates = getRandomCandidates(maxTasksPerStage, nodeMap, existingTasks);
+        NodeSelection randomNodeSelection = new RandomNodeSelection(nodeMap, includeCoordinator, minCandidates, maxTasksPerStage, existingTasks);
         Set<InternalNode> blockedExactNodes = new HashSet<>();
         boolean splitWaitingForAnyNode = false;
+        List<HostAddress> sortedCandidates = NodeSelector.sortedNodes(nodeMap);
         for (Split split : splits) {
-            randomCandidates.reset();
-
             List<InternalNode> candidateNodes;
-            if (!split.isRemotelyAccessible()) {
-                candidateNodes = selectExactNodes(nodeMap, split.getAddresses(), includeCoordinator);
-            }
-            else {
-                candidateNodes = selectNodes(minCandidates, randomCandidates);
+            switch (split.getNodeSelectionStrategy()) {
+                case HARD_AFFINITY:
+                    candidateNodes = selectExactNodes(nodeMap, split.getPreferredNodes(sortedCandidates), includeCoordinator);
+                    break;
+                case SOFT_AFFINITY:
+                    candidateNodes = convertToInternalNode(nodeMap, split.getPreferredNodes(sortedCandidates));
```

Comment:
Need to concat a list of random nodes to the end. The logic below `if (totalSplitCount < min && totalSplitCount < maxSplitsPerNode) {` needs to be adjusted accordingly so that we always pick the first two nodes unless they are busy. Otherwise, just select one from the concat'ed random nodes.